[MRG] text vectorizers should raise warnings when user params will be unused

getgaurav2 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
#14580 

#### What does this implement/fix? Explain your changes.
Add Warning and test cases for unused parameters for text vectorizers.

1.  Add  all warnings in a new method   - _warn_for_unused_params() and call this method in the fit() method for all vectorizers . 
2. Add Unit test case for all warning checks .
#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
Adding link to User guide for make_column_transformer

Commit:
Describe the arguments for *transformers.

Commit:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	sklearn/compose/_column_transformer.py

Commit:
#14580 - Add first warning for CountVectorizer for stop_words and analyzer pair

Commit:
removing spaces

Commit:
changing message

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

Commit:
#14580 - Add Warning for analyzer and stop word pair

Commit:
modifying warning text

Commit:
moving the code logic to fit

Commit:
bracket

Commit:
flake8 changes

Commit:
Adding Unit test case for Stop Word and analyzer != 'word

Commit:
flake8 changes

Commit:
more flake8

getgaurav2 commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -1105,7 +1105,7 @@ def test_tfidf_vectorizer_type(vectorizer_dtype, output_dtype,
     with pytest.warns(expected_warning_cls,
                       match=warning_msg_match) as record:
-            X_idf = vectorizer.fit_transform(X)
+        X_idf = vectorizer.fit_transform(X)
```
@jnothman - Can you please check this line in particular  . It was giving me a flake8 error so had to touch it to pass the tests.

getgaurav2 commented:
@jnothman  , @rth - can you pls. give any feedback on the test case ? 
I can add the other text vectorizers in the **same function**  and iterate over the objects of each  vectorizers  and check different  warnings msg  for different  parameter combinations 
OR 
 I can create one **unit test function  per vectorizer per  warning  .**



(py36) bash-3.2$ **pytest sklearn/feature_extraction/tests**
================================================================================ test session starts ================================================================================
platform darwin -- Python 3.6.8, pytest-5.0.1, py-1.8.0, pluggy-0.12.0
rootdir: /Users/XXXXX/Documents/GauravColumbiaUniv/sklearn-firstIssue, inifile: setup.cfg
collected 152 items

sklearn/feature_extraction/tests/test_dict_vectorizer.py ............................ [ 18%]
sklearn/feature_extraction/tests/test_feature_hasher.py .......... [ 25%]
sklearn/feature_extraction/tests/test_image.py .................... [ 38%]
**sklearn/feature_extraction/tests/test_text.py ................................................................x.........x..x..............x. [100%]**

================================================================= 148 passed, 4 xfailed, 6 warnings in 6.23 seconds =================================================================
(py36) bash-3.2$ **pytest -v sklearn/feature_extraction/tests/test_text.py::test_CountVectorizer_fit_warning**
================================================================================ test session starts ================================================================================
platform darwin -- Python 3.6.8, pytest-5.0.1, py-1.8.0, pluggy-0.12.0 -- /Users/XXXXX/GauravXXXX/Columbia/Anaconda/anaconda/envs/py36/bin/python
cachedir: .pytest_cache
rootdir: /Users/XXXXX/Documents/GauravColumbiaUniv/sklearn-firstIssue, inifile: setup.cfg
collected 1 item

**sklearn/feature_extraction/tests/test_text.py::test_CountVectorizer_fit_warning PASSED [100%]**

============================================================================= 1 passed in 0.35 seconds ==============================================================================

rth commented on sklearn/feature_extraction/text.py:
```diff
@@ -1091,6 +1091,9 @@ def fit(self, raw_documents, y=None):
         """
+        if self.analyzer != 'word':
+            warnings.warn("The parameter 'stop_words' will not be used"
+                          " since analyzer != 'word'")
```
This should only be raised if `stop_words is not None`

getgaurav2 replied:
done

rth commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -1288,3 +1288,13 @@ def analyzer(doc):
+
+
+def test_CountVectorizer_fit_warning():
+    CountVectorizerMsg = "The parameter 'stop_words' will not be used"
```
```suggestion
    msg = "The parameter 'stop_words' will not be used"
```

rth commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -1288,3 +1288,13 @@ def analyzer(doc):
+        vec.set_params(stop_words=["you've", "you", "you'll", 'AND'],
+                       analyzer='char')
+        assert_warns_message(UserWarning, CountVectorizerMsg, vec.fit,
+                             ['hello world'])
```
Better to use,
```py
with pytest.warns(UserWarning, match=msg):
    vec.fit(...)
```

getgaurav2 replied:
Done

rth commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -1288,3 +1288,13 @@ def analyzer(doc):
         Estimator(analyzer=analyzer, input='file').fit_transform([f])
+
+
+def test_CountVectorizer_fit_warning():
```
function names need to be all lowercase

This could also be,
```
@pytest.mark.parametrize("Vectorizer", [CountVectorizer, TfidfVectorizer, HashingVectorizer])
def test_unused_parameters_warn(Vectorizer):
    vect = Vectorizer()
```

getgaurav2 replied:
changed the function name for now ...will parameterize it later tonight

getgaurav2 replied:
Made the function parameterized . Have only included [CountVectorizer] for now. Will add the others as I add warnings for them  .

getgaurav2 committed:
Update sklearn/feature_extraction/tests/test_text.py

Co-Authored-By: Roman Yurchak <rth.yurchak@gmail.com>

Commit:
incorporating suggestions from @rth

Commit:
flake 8 suggestions

jnothman reviewed (commented):
Looking good

Commit:
Using pytest.warns and parameterizing  the test case .

getgaurav2 commented:
<img width="1275" alt="Screen Shot 2019-08-08 at 7 54 15 PM" src="https://user-images.githubusercontent.com/4666017/62745450-fc83fa80-ba18-11e9-875a-9599313f9052.png">

getgaurav2 reviewed (commented):
Adding only CountVectorizer for now . Will add others as I add warnings for them in the text.py file.

Commit:
Adding minor changes to trigger another build since a few test cases failed.

getgaurav2 commented:
@jnothman  , @rth  - can you please check why some of the tests are failing for this commit. 
Here is the  log  for one of them : 
**scikit-learn.scikit-learn (Linux py35_conda_openblas)** 
PackagesNotFoundError: The following packages are not available from current channels:

  - numpy=1.11.0
  - scipy=0.17.0
  - matplotlib=1.5.1
  - pillow=4.0.0

Commit:
Adding Warning and Unit test for analyzer and tokenizer pair

Commit:
Adding warning and unit test for tokenizer and token-pattern pair

thomasjpfan commented:
It is a unrelated issue. 

It should be fixed when https://github.com/scikit-learn/scikit-learn/pull/14619 get merged.

rth commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -1288,3 +1288,31 @@ def analyzer(doc):
+    with pytest.warns(UserWarning, match=msg_stop_words):
+        vect.fit(train_data)
+    vect.set_params(tokenizer=lambda s: s.split(),
+                    analyzer='char')
```
You are re-using parameters from previous modified instae, let's create a new instance at each time,
```py
vect = Vectorizer()
vect.set_params(...)
msg = '...'
with pytest.warns(UserWarning, match=msg):
   vect.fit(train_data)

etc. 
```
it's much easier to read if message is define the line above from where it's used.

getgaurav2 replied:
Done

rth commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -1288,3 +1288,31 @@ def analyzer(doc):
+
+
+@pytest.mark.parametrize(
+    'Vectorizer', [CountVectorizer]
```
If TfidfVectorizer and HashingVectorizer also pass this test (as they inherit from CountVectorizer) please add them here.

getgaurav2 replied:
Actually , I did try all three  earlier but  it failed .  The original post also mentioned
from sklearn.feature_extraction.text import **CountVectorizer**
so I stuck to that  
<img width="748" alt="Screen Shot 2019-08-10 at 10 26 05 AM" src="https://user-images.githubusercontent.com/4666017/62822967-483dbd80-bb59-11e9-8cd3-2e9296ab6bf9.png">

getgaurav2 replied:
Sure . Will try and get back to you .

rth reviewed (approved):
LGTM, apart for the two comments below.

Commit:
creating new instance of Vectorizer() for each test and reusing msg variable.

Commit:
Adding Warning and Unit test for callable(analyzer) and preprocessor pair
Adding Warning and Unit test for callable(analyzer) and ngram_range pair

Commit:
Modifying the documentation

getgaurav2 commented:
@rth  , @jnothman  - I have added warning and test cases for the following pairs : 

- stop_words is not used when analyzer != 'word'
- tokenizer is not used when analyzer != 'word'
- token_pattern is not used when tokenizer is not None
- preprocessor is not used  when callable(analyzer) -- changed user guide 
- ngram_range is not used when  callable(analyzer) -- changed user guide 
- **Ques** : token_pattern is not used when analyzer!= 'word' - this was not part of the original request  . Will this require a warning and test case as well ?

Commit:
Adding warning in the fit methods for HashingVectorizer and TfidfVectorizer

Commit:
change  documentation for CountVectorizer ,  HashingVectorizer for  preprocessor and ngram_range params )

jnothman commented:
Yes, add warnings for whichever cases you can identify. And yes, use a helper method, but don't use camel case to name it.

Commit:
Adding warning and unit test for token_pattern and analyzer != 'word'

jnothman commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -1288,3 +1288,66 @@ def analyzer(doc):
+    'Vectorizer',
+    [CountVectorizer, HashingVectorizer, TfidfVectorizer]
+)
+def test_unused_parameters_warn(Vectorizer):
```
It might also be neat if we parameterized the test with parameters like (unused_name, unused_value, overriding_name, overriding_value) and just match the part of the message that says "{unused_name}' will not be used.

getgaurav2 replied:
Done . Had to add a condition to check for ngram_range != (1, 1)  in the parameters and the warning checks . 
`        if (self.ngram_range != (1, 1) and self.ngram_range is not None
                and callable(self.analyzer)):
            warnings.warn("The parameter 'ngram_range' will not be used"
                          " since 'analyzer' is callable'")`

jnothman commented on sklearn/feature_extraction/text.py:
```diff
@@ -681,6 +683,30 @@ def fit(self, X, y=None):
 
         self._validate_params()
 
+        if self.analyzer != 'word' and self.stop_words is not None:
```
Can this just happen in _build_analyzer to avoid code duplication?

getgaurav2 replied:
@jnothman  -I have moved the analyzer related warnings inside the build_analyzer . Can  you please check if this is what you had in mind . Also , the  unit test for HashVectorizer fails   with this approach . Need to find the root cause for that .

getgaurav2 commented:
@jnothman - thanks for the review . I see #14619 is now merged . Could you also take a look at some of the checks that are failing

jnothman commented:
I've not looked at the logs here and can't right now, but if everything's failing it's usually due to a simple error in the pull request code

jnothman commented:
Go take a look at the logs and see if you can diagnose

getgaurav2 commented:
Okay . I will check the logs and try .
Thank you

Commit:
Adding _missing_param_warning(self)

Commit:
Adding analyzer!=word  warnings inside the  build_analyzer function

jnothman commented:
HashingVectorizer doesn't do much in fit. I suspect we should change it to
call _build_analyzer just as a smoke test.

Commit:
calling build_analyzer inside fit() method for Hashing vectorizer

jnothman commented on sklearn/feature_extraction/text.py:
```diff
@@ -714,7 +714,7 @@ def fit(self, X, y=None):
 
         self._validate_params()
-
+        self.build_analyzer()
```
Add a comment to say that this is used for further validation

getgaurav2 replied:
done

jnothman commented on sklearn/feature_extraction/text.py:
```diff
@@ -401,10 +401,28 @@ def build_analyzer(self):
         preprocess = self.build_preprocessor()
 
         if self.analyzer == 'char':
+            if self.stop_words is not None:
```
This shouldn't be duplicated, and it should also apply if analyzer is a callable

getgaurav2 replied:
Done .

jnothman reviewed (commented):
Thanks

Commit:
1. Further parameterizing  the unit test function
2. Adding condition for self.analyzer != 'word' or callable(self.analyzer) in build_analyzer

jnothman commented:
Tests are failing . Let us know if you need help

Commit:
Merge branch 'master' of https://github.com/getgaurav2/scikit-learn into 14580

Commit:
flake8 changes

Commit:
resolving merge conflicts

getgaurav2 commented:
> Tests are failing . Let us know if you need help

Thanks. A different unit test test_callable_analyzer_change_behavior seems to be failing ...I'll check the reason and get back to you .

Commit:
removing an extra warning added by mistake.

Commit:
commenting assert len(records) == 1 for the time

getgaurav2 commented:
> > Tests are failing . Let us know if you need help
> 
> Thanks. A different unit test **test_callable_analyzer_change_behavior** seems to be failing ...I'll check the reason and get back to you .

@jnothman  - Could you help me understand the reason for adding assert len(records) == 1 in test_callable_analyzer_change_behavior . This  condition is failing for HashingVectorizer ( screen shot below) 
<img width="1277" alt="Screen Shot 2019-08-24 at 10 56 27 AM" src="https://user-images.githubusercontent.com/4666017/63639178-6bd72c80-c65e-11e9-8f4d-3de2960e6630.png">

jnothman commented:
It's saying that HashingVectorizer is now issuing the ChangedBehaviorWarning twice. It should only issue it once. By putting build_analyzer in its fit, it might now be repeating that warning.

getgaurav2 commented:
> It's saying that HashingVectorizer is now issuing the ChangedBehaviorWarning twice. It should only issue it once. By putting build_analyzer in its fit, it might now be repeating that warning.

Yes , build_analyzer () is now being called inside fit() as well as transform() methods which is likely causing  the double warnings . Looking into it now .

Commit:
Taking the analyzer!= word validations outside of build_analyzer

jnothman commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -1266,6 +1266,7 @@ def test_callable_analyzer_change_behavior(Estimator, analyzer, input_type):
         with pytest.warns(ChangedBehaviorWarning, match=warn_msg) as records:
             Estimator(analyzer=analyzer, input=input_type).fit_transform(data)
     assert len(records) == 1
+    print(str(records[0]))
```
please avoid printing

getgaurav2 replied:
done

jnothman commented on sklearn/feature_extraction/text.py:
```diff
@@ -466,6 +468,32 @@ def _validate_params(self):
                 "lower boundary larger than the upper boundary."
                 % str(self.ngram_range))
 
+    def _missing_param_warning(self):
```
Rename to `_warn_for_unused_params`

getgaurav2 replied:
done

jnothman commented on sklearn/feature_extraction/text.py:
```diff
@@ -679,6 +709,8 @@ def fit(self, X, y=None):
                 "string object received.")
 
+        self._missing_param_warning()
+
```
rm blank line

getgaurav2 replied:
done

jnothman commented on sklearn/feature_extraction/text.py:
```diff
@@ -1091,6 +1125,8 @@ def fit(self, raw_documents, y=None):
         self
         """
+        self._missing_param_warning()
+
```
rm blank line

getgaurav2 replied:
done

jnothman reviewed (approved):
Otherwise LGTM. (Though I've not checked that the set of parameter pairs is complete.)

Please add an `Enhancement` entry to the change log at `doc/whats_new/v0.22.rst`. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`

jnothman commented:
Assuming you also think this is ready for merge, please change WIP -> MRG

Commit:
Rename the warning function from _missing_param_warning to _warn_for_unused_params
And remove blank lines

jnothman commented:
Please add a note in what's new as requested above. Thanks!

Commit:
Adding enhancement section for this pull request

getgaurav2 commented:
> Please add a note in what's new as requested above. Thanks!

Yup ..Was just working on it. Added now .

Commit:
minor fixes

getgaurav2 commented on doc/whats_new/v0.22.rst:
```diff
@@ -155,6 +163,10 @@ Changelog
   removed in v0.24. :pr:`14520` by
   :user:`Guillem G. Subies <guillemgsubies>`.
 
+- |Enhancement| :func:`_warn_for_unused_params` will give warning
```
Removed.

Commit:
modifying language

getgaurav2 commented on doc/whats_new/v0.22.rst:
```diff
@@ -146,6 +146,14 @@ Changelog
 :mod:`sklearn.feature_extraction`
 .................................
 
+- |Enhancement| :func:`feature_extraction.text._warn_for_unused_params` will
```
Got it ..Can you please check if the latest commit  makes more sense .

jnothman commented:
Makes more sense. We might yet find clearer words before release, but as far as I'm concerned we can merge. Thanks, @getgaurav2!
